main.go

```diff
@@ -0,0 +1,100 @@
+package main
+
+import (
+	"context"
+	"errors"
+	ginzap "github.com/gin-contrib/zap"
+	"github.com/gin-gonic/gin"
+	"go.uber.org/fx"
+	"go.uber.org/fx/fxevent"
+	"go.uber.org/zap"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"moul.io/zapgorm2"
+	"net/http"
+	"time"
+)
+
+func provideLogger() *zap.Logger {
+	logger, _ := zap.NewProduction()
+	return logger
+}
+
+func provideRouter(logger *zap.Logger) *gin.Engine {
+	r := gin.New()
+	r.Use(ginzap.Ginzap(logger, time.RFC3339, true))
+	r.Use(ginzap.RecoveryWithZap(logger, true))
+	return r
+}
+
+func invokeRoutes(r *gin.Engine, db *gorm.DB) {
+	r.GET("/", func(c *gin.Context) {
+		c.JSON(200, gin.H{"message": "hello"})
+	})
+	r.GET("/ping", func(c *gin.Context) {
+		var product Product
+		db.First(&product, 1)
+		c.JSON(200, product)
+	})
+}
+
+func invokeHttpServer(lc fx.Lifecycle, ginEngine *gin.Engine, logger *zap.Logger) {
```

Comment:
Like the routes, this instantiates the server so it needs to be a invoke rather than a `provider`